### PR TITLE
Device plugin

### PR DESCRIPTION
**Pull request by Eoghan1232**
This PR is a WIP, and is based on top of @bn222 2 PR's, which should be merged prior to this.

### CODE REVIEW

File: Dockerfile.daemon.rhel
Comment (Eoghan1232) [inline Dockerfile.daemon.rhel:2]: undo - local testing only

File: Dockerfile.daemon.rhel
Comment (Eoghan1232) [inline Dockerfile.daemon.rhel:20]: undo - local testing only

File: Dockerfile.rhel
Comment (Eoghan1232) [inline Dockerfile.rhel:2]: undo - local testing only

File: Dockerfile.rhel
Comment (Eoghan1232) [inline Dockerfile.rhel:22]: undo - local testing only

File: bindata/daemon/04.daemonset.yaml
Comment (Eoghan1232) [inline bindata/daemon/04.daemonset.yaml:26]: undo - local testing only

File: bindata/daemon/04.daemonset.yaml
Comment (Eoghan1232) [inline bindata/daemon/04.daemonset.yaml:39]: remove ``/``

File: config/dev/local-images.yaml
Comment (Eoghan1232) [inline config/dev/local-images.yaml:22]: undo - local testing only

File: config/dev/local-images.yaml
Comment (Eoghan1232) [inline config/dev/local-images.yaml:25]: undo - local testing only

File: daemon/device-plugin/deviceplugin.go
Comment (Eoghan1232) [inline daemon/device-plugin/deviceplugin.go:227]: pretty sure when I implemented DP before, these were required as I was implementing the interface...
so this is why these are here

File: daemon/device-plugin/types.go
Comment (Eoghan1232) [inline daemon/device-plugin/types.go:1]: remove header?

File: daemon/device-plugin/types.go
Comment (Eoghan1232) [inline daemon/device-plugin/types.go:23]: move to types.go ?

File: daemon/dpudaemon.go
Comment (Eoghan1232) [inline daemon/dpudaemon.go:11]: These imports are invalid - as were required for local testing.
once PR is merged, a second PR will be required to correct the imports

File: dpu.yaml
Comment (Eoghan1232) [inline dpu.yaml:1]: do we want this file or not? it's my own for locally testing, as there is no CR file on repo

File: go.mod
Comment (Eoghan1232) [inline go.mod:111]: TODO: remove after PR is merged

**Comment by Eoghan1232** [comment]
@bn222 please review as well. 
Please only look at latest commit, I kept all changes in one for you.

File: Dockerfile.daemon.rhel
Comment (Eoghan1232) [reply Dockerfile.daemon.rhel:2]: done

File: Dockerfile.rhel
Comment (Eoghan1232) [reply Dockerfile.rhel:22]: done

File: bindata/daemon/04.daemonset.yaml
Comment (Eoghan1232) [reply bindata/daemon/04.daemonset.yaml:26]: done - but the ``{{.DpuOperatorDaemonImage}}`` does not work, hence the local change

File: daemon/device-plugin/types.go
Comment (Eoghan1232) [reply daemon/device-plugin/types.go:23]: I will leave here for now.

File: daemon/device-plugin/deviceplugin.go
Comment (bn222) [inline daemon/device-plugin/deviceplugin.go:174]: what if we call `Start` more than once. Is that an issue?

File: daemon/dpudaemon.go
Comment (bn222) [inline daemon/dpudaemon.go:73]: we need wg and handle the case where one of the routines dies unrecoverably.

File: daemon/hostdaemon.go
Comment (bn222) [inline daemon/hostdaemon.go:185]: make sure to run go fmt before pushing

File: daemon/main.go
Comment (bn222) [inline daemon/main.go:37]: I like how you dependency-inject the dp.

File: dpu-api/device_service.proto
Comment (bn222) [inline dpu-api/device_service.proto:28]: Can you explain why we need map instead of `repeated`

File: dpu-api/gen/device_service.pb.go
Comment (bn222) [inline dpu-api/gen/device_service.pb.go:1]: Definately should commit this in a separate commit and need a `Makefile` rule

File: dpu-cni/example/cniserver_main.go
Comment (bn222) [inline dpu-cni/example/cniserver_main.go:7]: is this a `go fmt` change?

File: dpu.yaml
Comment (bn222) [reply dpu.yaml:1]: I didn't commit mine but we can add it in an examples diectory

File: go.mod
Comment (bn222) [inline go.mod:14]: We prefer to commit `go mod vendor` in the final commit in the PR, because if you `go mod vendor` and add it in 1 commit, it makes it hard to review (getting rid of `go mod vendor` won't happen any time soon)

File: daemon/device-plugin/deviceplugin.go
Comment (wizhaoredhat) [inline daemon/device-plugin/deviceplugin.go:96]: We should Get devices by through the "Resolve" API that would be implemented by the VSP.
So we should do  ports = Resolve(POD_UUID, VLAN ID (a.k.a chain ID) that would store this data in the VSP and return back the ports (APF interfaces) in the following struct : 
Name str;
PortType enum; (PortType can be NetDev or PCIEDev or VirtualIf - Standard Linux Interface Constructs)

This would be implemented in the NF CNIServer code. Let me know if you want me to do a prototype writeup for this.

File: daemon/device-plugin/deviceplugin.go
Comment (wizhaoredhat) [inline daemon/device-plugin/deviceplugin.go:98]: The device plugin's purpose should only make sure the DPU has enough resources to allocate a NF. It should not be tied to any interfaces. This should be instead taken care of the NF-CNI. If there is no "port" resources left then the NF-CNI should fail and the NF would be in "ErrorCreating" state.

**Review by wizhaoredhat — CHANGES_REQUESTED** [review]
Me and @bn222 Talked about the ideal architecture of the NF Device Plugin and NF CNI and how interfaces should be exposed. This was done today and perhaps you did not get an update on this.

File: daemon/device-plugin/deviceplugin.go
Comment (wizhaoredhat) [inline daemon/device-plugin/deviceplugin.go:1]: rename to NF device plugin

File: bindata/daemon/04.daemonset.yaml
Comment (Eoghan1232) [reply bindata/daemon/04.daemonset.yaml:49]: It's the kubelet device plugin socket directory

File: daemon/device-plugin/deviceplugin.go
Comment (Eoghan1232) [reply daemon/device-plugin/deviceplugin.go:33]: yes can rename

File: daemon/device-plugin/deviceplugin.go
Comment (Eoghan1232) [reply daemon/device-plugin/deviceplugin.go:51]: the device requires a state, for now, it will always return healthy

File: daemon/device-plugin/deviceplugin.go
Comment (Eoghan1232) [reply daemon/device-plugin/deviceplugin.go:96]: understood, APF can be renamed, this is leftover from my other implementation

File: daemon/device-plugin/deviceplugin.go
Comment (Eoghan1232) [reply daemon/device-plugin/deviceplugin.go:128]: forgot to move it up, think I was moving stuff around and it got forgotten

File: daemon/device-plugin/deviceplugin.go
Comment (Eoghan1232) [reply daemon/device-plugin/deviceplugin.go:138]: a long time, this path is still used for storing unix socket files for device plugins.
we also still use this in SRIOV DP

File: daemon/device-plugin/deviceplugin.go
Comment (Eoghan1232) [reply daemon/device-plugin/deviceplugin.go:155]: I will take a look

File: daemon/device-plugin/deviceplugin.go
Comment (Eoghan1232) [reply daemon/device-plugin/deviceplugin.go:64]: ListAndWatch is an implemented function, I can change what's inside, but not the function headings. 

to clarify - you want the ListAndWatch to make the call to another function which will return the list of devices ? 
so instead of the devices being populated on startup, you want it routinely check?

File: daemon/device-plugin/deviceplugin.go
Comment (Eoghan1232) [reply daemon/device-plugin/deviceplugin.go:174]: I think it would fail to register, not sure. 
could be fine...

File: daemon/main.go
Comment (Eoghan1232) [reply daemon/main.go:37]: it was a necessity :)

File: dpu-api/device_service.proto
Comment (Eoghan1232) [reply dpu-api/device_service.proto:28]: good question - this is just how I logically did it
we did not define a proto definition, so this is what I went with.

would you prefer if it was ``repeated`` ?

File: dpu-cni/example/cniserver_main.go
Comment (Eoghan1232) [reply dpu-cni/example/cniserver_main.go:7]: I believe so

File: daemon/device-plugin/deviceplugin.go
Comment (Eoghan1232) [reply daemon/device-plugin/deviceplugin.go:51]: right - but this is more post demo I think.
right now, it will always report true, until this API or check can be defined

File: daemon/device-plugin/deviceplugin.go
Comment (Eoghan1232) [reply daemon/device-plugin/deviceplugin.go:35]: why? a device represents a single device, NF would represent 2 ? 
or am I missing something?

File: daemon/device-plugin/deviceplugin.go
Comment (Eoghan1232) [reply daemon/device-plugin/deviceplugin.go:98]: why?
standard, the DP would put the resource as an env into the pod, it's only tied to a specific APF interface, as we are using the interface name as a way to move the device from host to pod namespace right?
I am confused, when you say port, you mean the interface moved inside the pod right? port = APF = ?

File: daemon/device-plugin/deviceplugin.go
Comment (Eoghan1232) [reply daemon/device-plugin/deviceplugin.go:1]: sure

File: dpu.yaml
Comment (Eoghan1232) [reply dpu.yaml:1]: ok I will move it. 
I will also create a "host" one too

File: daemon/device-plugin/deviceplugin.go
Comment (Eoghan1232) [reply daemon/device-plugin/deviceplugin.go:51]: ListAndWatch is continuously called afaik, if something were changed, then it gets updated.
not leftover

File: daemon/device-plugin/deviceplugin.go
Comment (Eoghan1232) [reply daemon/device-plugin/deviceplugin.go:96]: I will rename to "NF-DEV" for now

File: daemon/device-plugin/deviceplugin.go
Comment (Eoghan1232) [reply daemon/device-plugin/deviceplugin.go:155]: done - implemented a retry function

File: daemon/device-plugin/deviceplugin.go
Comment (Eoghan1232) [reply daemon/device-plugin/deviceplugin.go:174]: if unix socket is not cleaned up - then it will fail, as address will be in use. 
I tested this. 
I have a ``cleanup`` function there, just it needs to be called.

File: daemon/device-plugin/deviceplugin.go
Comment (Eoghan1232) [reply daemon/device-plugin/deviceplugin.go:64]: I have not reimplemented this yet - I would like a sync to discuss flow before changes, in case I misunderstood.

File: daemon/device-plugin/deviceplugin.go
Comment (Eoghan1232) [reply daemon/device-plugin/deviceplugin.go:170]: currently don't have any call to cleanup - need to check best time to do it

Comment (Eoghan1232) [reply daemon/device-plugin/deviceplugin.go:170]: I have added this in, currently testing - will commit in a moment.

Comment (Eoghan1232) [reply daemon/device-plugin/deviceplugin.go:170]: fixed - ensure clean state at start

File: dpu-api/gen/device_service.pb.go
Comment (Eoghan1232) [reply dpu-api/gen/device_service.pb.go:1]: added to my to-do

File: daemon/device-plugin/deviceplugin.go
Comment (bn222) [reply daemon/device-plugin/deviceplugin.go]: +1, post demo is ok.

File: daemon/device-plugin/deviceplugin.go
Comment (bn222) [reply daemon/device-plugin/deviceplugin.go:64]:
```
func (sm *apfManager) ListAndWatch(empty *pluginapi.Empty, stream pluginapi.DevicePlugin_ListAndWatchServer) error {
	changed := true
	for {
             // call helper here that does the heavy lifting.
             // take result and pass it into stream
        }
}
```

That's all.

File: daemon/device-plugin/deviceplugin.go
Comment (bn222) [reply daemon/device-plugin/deviceplugin.go]: Let's park this for now. We will revisit it in 3 weeks

File: daemon/device-plugin/deviceplugin.go
Comment (bn222) [reply daemon/device-plugin/deviceplugin.go:98]: Same comment as above, I'm ok to tackle this in May.

File: daemon/device-plugin/deviceplugin.go
Comment (bn222) [reply daemon/device-plugin/deviceplugin.go:137]: yes, I understand that part. We are good as long as it's not getting removed between now and 2 months. The question is: will it remain in a deprecated state for 2 months and get removed after or is there a plan in k8s to remove it earlier? A quick check would answer that question.

File: daemon/device-plugin/deviceplugin.go
Comment (bn222) [reply daemon/device-plugin/deviceplugin.go:170]: ty. this part lgtm.

File: daemon/dpudaemon.go
Comment (bn222) [reply daemon/dpudaemon.go:73]: :) ok.

File: dpu-api/device_service.proto
Comment (bn222) [reply dpu-api/device_service.proto:28]: It's not that I prefer `map` or `repeated`. Rather, I want to check if there is a real reason for map? Do we have keys + values? Or is it "just a list of values".

File: daemon/device-plugin/deviceplugin.go
Comment (Eoghan1232) [reply daemon/device-plugin/deviceplugin.go:137]: https://kubernetes.io/docs/concepts/extend-kubernetes/compute-storage-net/device-plugins/#device-plugin-implementation

not sure if it is deprecated actually - that was leftover from when I did it.... 
it's still on the official docs as the path to use

File: dpu-api/device_service.proto
Comment (Eoghan1232) [reply dpu-api/device_service.proto:28]: it's key + values, as the device could have more info than just the ID.

File: daemon/device-plugin/deviceplugin.go
Comment (Eoghan1232) [reply daemon/device-plugin/deviceplugin.go:56]: done

File: daemon/device-plugin/deviceplugin.go
Comment (wizhaoredhat) [reply daemon/device-plugin/deviceplugin.go]: Agreed

File: daemon/device-plugin/deviceplugin.go
Comment (wizhaoredhat) [reply daemon/device-plugin/deviceplugin.go:103]: Agreed

File: daemon/device-plugin/types.go
Comment (wizhaoredhat) [inline daemon/device-plugin/types.go]: Also remove license blob here.

File: Dockerfile.daemon.rhel
Comment (Eoghan1232) [reply Dockerfile.daemon.rhel]: lets address in separate PR - that is a local dev issue :)

File: config/dev/local-images.yaml
Comment (Eoghan1232) [reply config/dev/local-images.yaml]: old commit right ? should not appear in latest - also this entire PR has to be rebased once merge from Bal

File: daemon/device-plugin/deviceplugin.go
Comment (Eoghan1232) [reply daemon/device-plugin/deviceplugin.go:28]: demo? :p 

okay will change.

File: daemon/device-plugin/types.go
Comment (Eoghan1232) [reply daemon/device-plugin/types.go]: think already removed from up to date commit? let me confirm

Comment (Eoghan1232) [reply daemon/device-plugin/types.go]: confirmed - not in latest commit. a squash later will clean the history

**Pull request closed without merging**
